python-lib/text_extraction/__init__.py

```diff
@@ -54,4 +54,108 @@ def extract_text_content(file_bytes, extension, with_pandoc):
+                tmp.write(file_bytes)
+                markdown = pypandoc.convert_file(tmp.name, to="md", format=extension)
+        except Exception as e:
+            raise ValueError("Cannot convert file into markdown using pandoc because: {}".format(e))
+    
+        chunks = extract_markdown_chunks(markdown, filename)
+        return chunks
+
+
+def extract_markdown_chunks(markdown, filename):
+    # Code largely inspired by `langchain.text_splitter.HTMLHeaderTextSplitter`
+    lines = markdown.split("\n")
+    # Final output
+    lines_with_metadata = []
+    # text and metadata of the chunk currently being processed
+    current_text = []
+    current_metadata = {}
+    # Keep track of the nested header structure
+    # header_stack: List[Dict[str, Union[int, str]]] = []
+    header_stack = []
+    initial_metadata = {}
+
+    for line in lines:
+        stripped_line = line.strip()
+
+        # Check each line against each of the header types (e.g., #, ##), header_level is the number of '#'
+        for header_level in range(1, 4):
+            # Check if line starts with a header that we intend to split on
+            if stripped_line.startswith("#" * header_level) and (
+                # Header with no text OR header is followed by space are valid conditions that sep is being used a header
+                len(stripped_line) == header_level or stripped_line[header_level] == " "
+            ):
+
+                # Pop out headers of lower or same level from the stack
+                while header_stack and header_stack[-1]["level"] >= header_level:
+                    # We have encountered a new header at the same or higher level
+                    popped_header = header_stack.pop()
+                    # Clear the metadata for the popped header in initial_metadata
+                    if popped_header["level"] in initial_metadata:
+                        initial_metadata.pop(popped_header["level"])
+
+                # Push the current header to the stack
+                header = {
+                    "level": header_level,
+                    "data": stripped_line[header_level :].strip(),
+                }
+                header_stack.append(header)
+                # Update initial_metadata with the current header
+                initial_metadata[header_level] = header["data"]
+
+                # Add the previous line to the lines_with_metadata only if current_text is not empty
+                if current_text:
+                    lines_with_metadata.append(
+                        {
+                            "text": "\n".join(current_text),
+                            "metadata": current_metadata.copy(),
+                        }
+                    )
+                    current_text.clear()
+
+                break
+        else:
+            if stripped_line:
+                current_text.append(stripped_line)
+            elif current_text:
+                lines_with_metadata.append(
```

Comment:
Why do we only append to lines with metadata if the stripped line is empty ?

Reply:
This part of the code is to put different paragraphs of the same section (so separated with 2 new lines) in different rows (but with the same header), do we want that ?
Lines are appended to `lines_with_metadata` in the for loop when a new header is found

Reply:
explained better in [eb74caa](https://github.com/dataiku/dss-plugin-tesseract-ocr/pull/73/commits/eb74caa6e97e67b000bfd008b44e5a1c5984f900)

Reply:
Actually I don't think we want to separate paragraphs from the same header.
Let's start with the "simple" approach that consists in only chunking by header.

Btw which subheader level is the minimum level at which we define chunks ?

Reply:
stop separating paragrpahs in [fb16277](https://github.com/dataiku/dss-plugin-tesseract-ocr/pull/73/commits/fb16277d1efdbc45e474d8215562b8f4ede55f54)

the min subheader level considered is h4 (or ####)

Reply:
We should add this as a comment or even explain it in the inline help ?